src/python/WMComponent/WorkflowUpdater/SiteListPoller.py

```diff
@@ -0,0 +1,152 @@
+        """
+        Perform the following logic:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream wmstats server
+        - update site lists of active workflows
+        - push new specs to the agent local WorkQueue and update pickle spec file
+
+        :return: none
+        """
+        # get list of active workflows from the agent, the returned dict
+        # is composed by workflow names and associated pickle file (data comes from WMBS)
+        wmaSpecs = self.getActiveWorkflows()
+        wflows = wmaSpecs.keys()
+
+        # obtain workflow records from wmstats server
+        wdict = self.wmstatsDict(wflows)
+
+        # iterate over the list of active workflows which is smaller than list from wmstats
+        for wflow in wflows:
+            siteWhiteList = wdict[wflow]['SiteWhitelist']
+            siteBlackList = wdict[wflow]['SiteBlacklist']
+
+            # get the name of pkl file from wma spec
+            pklFileName = wmaSpecs[wflow]
+
+            # create wrapper helper and load pickle file
+            wHelper = WMWorkloadHelper()
+            wHelper.load(pklFileName)
+
+            # extract from pickle spec both white and black site lists and compare them
+            # to one we received from upstream service (ReqMgr2)
+            wmaWhiteList = wHelper.getSiteWhiteList()
+            wmaBlackList = wHelper.getSiteBlackList()
+            if set(wmaWhiteList) != set(siteWhiteList) or set(wmaBlackList) != set(siteBlackList):
+                self.logger.info("Updating %s:", wflow)
+                self.logger.info("  siteWhiteList %s => %s", wmaWhiteList, siteWhiteList)
+                self.logger.info("  siteBlackList %s => %s", wmaBlackList, siteBlackList)
+                try:
+                    # update local WorkQueue first
+                    self.localWQ.updateSiteLists(wflow, siteWhiteList, siteBlackList)
+                    self.logger.info("successfully updated workqueue elements for workflow %s", wflow)
+                except Exception as ex:
+                    logging.exception("Unexpected exception while updating elements in local workqueue Details:\n%s", str(ex))
+                    continue
+
+                # update workload only if we updated local WorkQueue
+                # update site white/black lists together
+                if set(wmaWhiteList) != set(siteWhiteList):
+                    self.logger.info("updating site white list for workflow %s", wflow)
+                    wHelper.setWhitelist(siteWhiteList)
+                if set(wmaBlackList) != set(siteBlackList):
+                    self.logger.info("updating site black list for workflow %s", wflow)
+                    wHelper.setBlacklist(siteBlackList)
+
+                try:
+                    # persist the spec in local CouchDB
+                    self.logger.info("Updating %s with new site lists for %s", self.localCouchUrl, wflow)
```

Comment:
Given that the pickle file is loaded from `pklFileName`, we might want to rephrase this record. Either just say we are updating it for workflow BLAH, or say that we are updating the pickle file at location XYZ.

Reply:
ok